File: 004.c
// 906609
#include <stdio.h>

int is_palindrome(int n)
{
    int reverse = 0;
    int n0 = n;
    while (n0 != 0) {
        reverse = (reverse*10) + (n0%10);
        n0 /= 10;
    }

    return reverse == n;
}

int main()
{
    int largest = 999;
    int smallest = 100;
    int max = largest;

    for (int i = largest; i >= smallest; i--) {
        for (int j = i; j >= smallest; j--) {
            int mul = i*j;
            if (is_palindrome(mul) && mul > max) {
                max = mul;
            }
        }
    }

    printf("%d\n", max);

    return 0;
}

File: zig/004.zig
const std = @import("std");
const stdout = std.io.getStdOut().writer();

fn is_palindrome(n: u64) bool {
    var reverse: u64 = 0;
    var n0: u64 = n;
    while (n0 != 0) {
        reverse = (reverse * 10) + (@mod(n0, 10));
        n0 /= 10;
    }

    return reverse == n;
}

fn max_list(list: std.ArrayList(u64)) u64 {
    var m: u64 = 0;
    for (list.items) |item| {
        m = @max(m, item);
    }

    return m;
}

fn solve(start: u64, end: u64) anyerror!u64 {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    var palindromes = std.ArrayList(u64).init(allocator);
    defer palindromes.deinit();

    for (start..end) |i| {
        for (start..end) |j| {
            var prod: usize = i * j;
            if (is_palindrome(prod)) {
                try palindromes.append(prod);
            }
        }
    }

    const max = max_list(palindromes);
    return max;
}

pub fn main() !void {
    const max = solve(100, 999);
    try stdout.print("{!}\n", .{max});
}
